comid/integregisters.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2024 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+package comid
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/veraison/swid"
+)
+
+// IRegisterIndex is the interface to hold register index
+type IRegisterIndex interface{}
+
+// IntegReg holds the Integrity Registers
+type IntegReg struct {
+	Register map[IRegisterIndex]Digests
+}
+
+func NewIntegRegister() (*IntegReg, error) {
+	reg := make(map[IRegisterIndex]Digests)
+	return &IntegReg{Register: reg}, nil
+}
+
+// AddDigests allows inserting a series of Digests at a specific RegisterIndex
+func (i *IntegReg) AddDigests(index IRegisterIndex, digests Digests) error {
+	if len(digests) == 0 {
+		return fmt.Errorf("no digests to add")
+	}
+	for _, digest := range digests {
+		if err := i.AddDigest(index, digest); err != nil {
+			return fmt.Errorf("unable to add Digest: %w", err)
+		}
+	}
+	return nil
+}
+
+// AddDigest allows inserting a Digest at a specific RegisterIndex
+func (i *IntegReg) AddDigest(index IRegisterIndex, digest swid.HashEntry) error {
+
+	if i.Register == nil {
+		return fmt.Errorf("no register to add digest")
+	}
+	switch t := index.(type) {
+	case string, uint, uint64:
+		i.Register[t] = append(i.Register[t], digest)
+	default:
+		return fmt.Errorf("unexpected type for index: %T", t)
+	}
+	return nil
+}
+
+// RemoveIndex removes an Index from the Register
+func (i *IntegReg) RemoveIndex(index IRegisterIndex) error {
+	if i.Register == nil {
+		return fmt.Errorf("cannot remove index, non existent register")
+	}
+	switch t := index.(type) {
+	case uint, uint64, string:
+		delete(i.Register, index)
+	default:
+		return fmt.Errorf("unknown type %T for index-type-choice", t)
+	}
+	return nil
+}
+
+func (i IntegReg) MarshalCBOR() ([]byte, error) {
+	return em.Marshal(i.Register)
+}
+
+func (i *IntegReg) UnMarshalCBOR(data []byte) error {
+	return dm.Unmarshal(data, &i.Register)
+}
+
+func (i IntegReg) ToJSON() ([]byte, error) {
```

Comment:
If I understand correctly, this mashals the key as a tnv JSON object, which is
then marshalled as string so that it can be used as a JSON key?
E.g.

```json
{
	"{\"type\": \"uint\", \"value\": 7}": [... encoded digests ...]
}
```

This double marshalling is weird and will resuilt in JSON text that is
difficult to prase visually. Why not move the key type to be part of the value,
so if would look something like

```json
{
	"7": {
		"key_type": "uint",
		"value": [... encoded digests ...]
	}
}
```

Reply:
Hmmm, I am not sure this representation accurately describes the `key type` though!

If I am not mis-taken: Here you are representing the `key`  as "7" in JSON which is a string representation

Reply:
To illustrate,

```json
{
	"7": {
		"key_type": "uint",
                [ ... ]
	}
}
```
Means the key is `7` (integer) 
```json
{
	"7": {
		"key_type": "text",
                [ ... ]
	}
}
```
means the key is `"7"` (string).

Reply:
Let me check, if I can improve the existing test-vector, that provides the required simplification without losing the `homogeneity`  offered by the `tnv` pattern we have been using consistently throughout this repository!

Reply:
In either case, you're breaking away form the `tnv` pattern, as it only works for JSON values -- that encodes a value a JSON object, which cannot be a key. In the original approach, you're then encoding the  `tnv` JSON object as a string, so it's no longer a `tnv` object.

Reply:
Done as required, other alternatives did not render very cleanly so implemented as above!